core/src/main/java/org/elasticsearch/threadpool/ThreadPool.java

```diff
@@ -1005,4 +1011,82 @@ public ThreadContext getThreadContext() {
         return threadContext;
     }
 
+    /**
+     * This interface represents an object whose execution may be cancelled during runtime.
+     */
+    public interface Cancellable {
+
+        /**
+         * Cancel the execution of this object. This method is idempotent.
+         */
+        void cancel();
+
+        /**
+         * Check if the execution has been cancelled
+         * @return true if cancelled
+         */
+        boolean isCancelled();
+    }
+
+    static class ReschedulingRunnable extends AbstractRunnable {
+
+        private final Runnable runnable;
+        private final TimeValue interval;
+        private final String name;
+        private final ThreadPool threadPool;
+
+        private volatile boolean run = true;
+
+        ReschedulingRunnable(Runnable runnable, TimeValue interval, String name, ThreadPool threadPool) {
+            this.runnable = runnable;
+            this.interval = interval;
+            this.name = name;
+            this.threadPool = threadPool;
+        }
+
+        Cancellable scheduleFirstRun() {
+            if (run == false) {
+                throw new IllegalStateException("this runnable should not be reused after cancelling");
+            }
+            threadPool.schedule(interval, name, this);
+            return new Cancellable() {
+                @Override
+                public void cancel() {
+                    run = false;
+                }
+
+                @Override
+                public boolean isCancelled() {
+                    return run == false;
+                }
+            };
+        }
+
+        @Override
+        public void doRun() {
+            if (run) {
+                runnable.run();
+            }
+        }
+
+        @Override
+        public void onFailure(Throwable t) {
+            // if a runnable throws, the default behavior of ScheduledThreadPoolExecutor is to stop execution
+            run = false;
+            threadPool.logger.warn("failed to run scheduled task [{}] on thread pool [{}]", runnable.toString(), name);
+        }
+
+        @Override
+        public void onAfter() {
+            if (run) {
+                threadPool.schedule(interval, name, this);
+            }
+        }
+    }
+
+    public static boolean assertNotScheduleThread() {
+        assert Thread.currentThread().getName().contains("scheduler") == false :
+            "Expected current thread [" + Thread.currentThread() + "] to not be the scheduler thread. Reason: " + "[Blocking operation]";
```

Comment:
strictly speaking, we can log the reason here as we don't know why people called us. We need to either pass a parameter (reason) or remove the last part.